File: 0148-sort-list/0148-sort-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeList(ListNode* list1, ListNode* list2){
        if(list1==nullptr)  return list2;
        if(list2==nullptr)  return list1;
        if(list1->val<list2->val){
            list1->next = mergeList(list1->next, list2);
            return list1;
        }else{
            list2->next = mergeList(list1, list2->next);
            return list2;
        }
    }

    ListNode* sortList(ListNode* left) {
        if(left==nullptr)   return left;
        if(left!=nullptr&&left->next==nullptr) return left;

        ListNode *middle=left, *fast=left->next;
        while(fast!=nullptr&&fast->next!=nullptr){
            middle=middle->next;
            fast=fast->next->next;
        }
        ListNode* right = middle->next;
        middle->next=nullptr;

        left = sortList(left);
        right = sortList(right);
        return mergeList(left, right);
    }
};

File: 0148-sort-list/0148-sort-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeList(self, left, right):
        if not left and right:  return right
        if not right and left:  return left

        dummy = ListNode(0)
        answer = dummy
        while left or right:
            left_val = left.val if left else float('inf')
            right_val = right.val if right else float('inf')
            if left_val < right_val:
                dummy.next = ListNode(left_val)
                left = left.next
            else:
                dummy.next = ListNode(right_val)
                right = right.next
            dummy = dummy.next

        return answer.next

    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return 
            
        if not head.next:
            return head
        
        # Find the middle nod
        mid = head
        fast = head.next
        while fast and fast.next:
            mid = mid.next
            fast = fast.next.next
        
        temp = mid.next
        mid.next = None
        
        # Divide and Conquer
        left = self.sortList(head)
        right = self.sortList(temp)

        return self.mergeList(left, right)

